### DISCUSSION TITLE
Need Help Creating an Elastic Stomach Model using BREP File and Gmsh in generating 3D mesh

### DISCUSSION THREAD

**Discussion by parabolang** [sofa-framework/sofa#4241] [Build / Config / Environment] [answered]
I am attempting to create an elastic stomach model for simulation using the SOFA framework. I have obtained an obj. file of the stomach, which I have successfully imported into SOFA to create a rigid stomach. However, my goal is to make the stomach model elastic.

To achieve this, I have been using FreeCAD to transform the obj. file into a BREP (Boundary Representation) file. My next step is to generate a 3D mesh using Gmsh from this BREP file, but I have encountered difficulties. When I attempt to use Gmsh, I can only generate a triangular mesh for the 2D surface, rather than a volume mesh.

Interestingly, when I create a simple cube in FreeCAD and export it as a BREP file, Gmsh is able to detect the volume and generate a tetrahedron mesh for the cube. However, this does not seem to work with the BREP file from the stomach model.

Is there a specific reason why Gmsh is unable to detect the volume for the BREP file from the stomach model? And how can I successfully generate an elastic tissue representation for the stomach model using the BREP file and Gmsh?

Any insights, suggestions, or alternative approaches would be greatly appreciated. Thank you in advance for your help!

Attached a BREP file,
[stoobj-stoobj001.zip](https://github.com/sofaframework/sofa/files/12999271/stoobj-stoobj001.zip)

**Comment by bakpaul** [comment, marked as answer]
Hi ! 

I am sorry but we cannot offer you support on Gmsh, this may get more response on their forum. 
But for you to know, you could already use your surface mesh in sofa. Using `SparseGrid` object, you can mesh automatically the volume represented by your obj into hexahedron. You could then transform it to tetrahedrons by using a `Hexa2TetraTopologicalMapping`. Here is a small example doing so. But at the end you'll have a plain organ, non hollow. 

``` xml
<Node name="RegularGrid">
	<SparseGrid name="HexaTop" cellWidth="20.0" fileTopology="yourSurfaceMesh.obj" />
	<TetrahedronSetTopologyContainer name="TetraContainer" />
	<TetrahedronSetTopologyModifier name="Modifier7" />
	<Hexa2TetraTopologicalMapping input="@HexaTop" output="@TetraContainer" swapping="false" />
</Node>
```

And also, you can use the CGal plugin available in Sofa to mesh your volume into tetrahedra. Again, here is a little example of this : 
``` xml
<Node name="root" dt="0.025" gravity="0 0 0" >
<RequiredPlugin pluginName="CGALPlugin"/>
<MeshObjLoader name="hull" filename="yourSurfaceMesh.obj" /> 
<MeshTopology src="@hull"/>

<MeshGenerationFromPolyhedron name="generator"
    inputPoints="@hull.position" inputTriangles="@hull.triangles" inputQuads="@hull.quads"
    cellSize="0.65" facetAngle="10"
    facetSize="0.3" cellRatio="2"
    facetApproximation="1" ordering="1"
    perturb="false" odt="false" lloyd="true" lloyd_max_it="5"/>

<Node name="Generated" >
    <TetrahedronSetTopologyContainer name="Container"  points="@../generator.outputPoints" tetras="@../generator.outputTetras"/>

    <MechanicalObject name="mstate" />
    <MeshExporter name='exporterA' format='vtk' printLog='true' filename='yourVolumicMesh' exportAtEnd='true' />
</Node>
</Node>  
```

**Reply by parabolang** [reply]
Thank you for your prompt response and your continued support. I'm glad to report that I have successfully implemented the first method you suggested, and it's working as expected "an elastic stomach". 

However, as I ventured into the second method, I encountered some challenges. It seems that I need to make further adjustments to various parameters. When I attempted to execute the second method, it didn't generate the expected VTK file, and the terminal abruptly terminated with a "Segmentation fault (core dumped)" error. I've included some of the processing scripts from the terminal for your reference. 

[WARNING] [Node(root)] An animation loop is required, but has not been found. Add an animation loop in the root of your scene to fix this warning. The list of available animation loop components is: [ConstraintAnimationLoop, DefaultAnimationLoop, FreeMotionAnimationLoop, MultiStepAnimationLoop, MultiTagAnimationLoop]. A component of type DefaultAnimationLoop will be automatically added for you.
Inserting protection balls...
  refine_balls = true
  min_balls_radius = 0
  min_balls_weight = 0
insert_corners() done. Nb of points in triangulation: 0
insert_balls_on_edges() done. Nb of points in triangulation: 0
refine_balls() done. Nb of points in triangulation: 0

construct initial points:
8/8 initial point(s) found...
Start surface scan...Scanning triangulation for bad facets (sequential) - number of finite facets = 31...
Number of bad facets: 8
scanning edges (lazy)
scanning vertices (lazy)
end scan. [Bad facets:8]

Refining Surface...
Legend of the following line: (#vertices,#steps,#facets to refine,#tets to refine)
(3890,3881,0,0) (1657.9 vertices/s))))
Total refining surface time: 2.34154s

Start volume scan...Scanning triangulation for bad cells (sequential)... 26251 cells scanned, done.
Number of bad cells: 12668
end scan. [Bad tets:12668]

Refining...
Legend of the following line: (#vertices,#steps,#facets to refine,#tets to refine)
(4060,169,0,12968) (954.2 vertices/s)

SIG 11 - SIGSEGV: segfault
...
Segmentation fault (core dumped)

**Reply by bakpaul** [reply]
To be clear, this method will give you the same behavior of the `SparseGrid` one  if you only use the surface mesh as input. It will mesh the entirety of the volume, the only difference will be that your mesh will follow the topology more. 

Be careful of the size given in the parameters `cellSize` and `facetSize`, this should be updated to fit your mesh size. (If your stomach is expressed in millimeter and is 200mm wide, then depending in the number of element you want, you should set these parameters accordingly, let's say 4 for approximately 50 element along the width). 

Moreover, the input polyhedron's quality will highly influence the generated mesh quality and the algorithm stability. For instance, I am not sure of the behavior if your polyhedron is not closed... 

If you only want to simulatet he surface you can either :
- Use shell elements
- Mesh the surface with really fine tetrahedrons using a polyhedron of the outer and the inner surface of the stomach, or there is also a way to mesh a binary image using `MeshGenerationFromImage`. I've used it a long time ago, but I couldn't find one of my former scene. But you can find an example in here `applications/plugins/CGALPlugin/scenes/MeshGenerationFromImage.scn`.

**Reply by parabolang** [reply]
Thanks for your prompt responses! The suggestions really helpful for me to generate mesh by using surface mesh.
